### PR TITLE
Introduce version and fix various bugs

### LINKED ISSUE #92: `JALR` causes wrong behavior on AsmMachine when `rs1` and `rd` use the same register.
**Describe the bug**

The following instruction will cause different behavior on DefaultMachine(rust) and AsmMachine(asm). 

``` asm
   1021a:	00000097          	auipc	ra,0x0
   1021e:	fd4080e7          	jalr	-44(ra) # 0x101ee
```

Let's focus on the instruction `1021e`. We expect this instruction to jump to `rs1`(which is `ra`) - `14` and set the `pc + 4` to the `rd`(which is `ra`) register. The DefaultMachine handles this as expected. But unfortunately, the AsmMachine set the registers in a wrong ordering:

https://github.com/nervosnetwork/ckb-vm/blob/develop/src/machine/asm/execute.S#L441

AsmMachine set `rd` first, then read from `rs1`, when the `rs1` and `rd` use the different registers this behavior is correct, but if the `rs1` and `rd` use the same register then the `WRITE_RD` operation overwrite the value of `rs1`.

**How to fix**

To simply fix this problem, we can move the `WRITE_RD` to the after of the  `REGISTER_ADDRESS(RS1)`.

The change may cause the network consensus split. We need to wait for a hard-fork to apply it.

### PR DESCRIPTION
**Pull request by xxuejie**
Since one major use case of CKB VM is on blockchains, we cannot just
change the code in case of bugs. We have to wait for the next fork
event, only then can we upgrade the code. Hence we are introducing a
concept of versions in the VM, whenever we fix a bug that might alter
behavior, we will insert checking code to only include the bug fix in
the next VM version. This way blockchains can expect consistent VM
behavior when they lock specific VM version.

This PR also provides VM version 1 which fixes #92 #97 and #98

### CODE REVIEW

File: definitions/src/instructions.rs [hunk 1]
```diff
@@ -24,141 +24,250 @@ pub type InstructionOpcode = u8;
...
+pub const OP_SRAIW: InstructionOpcode = 56;
+pub const OP_SRAW: InstructionOpcode = 57;
+pub const OP_SRL: InstructionOpcode = 58;
+pub const OP_SRLI: InstructionOpcode = 59;
+pub const OP_SRLIW: InstructionOpcode = 60;
+pub const OP_SRLW: InstructionOpcode = 61;
+pub const OP_SUB: InstructionOpcode = 62;
+pub const OP_SUBW: InstructionOpcode = 63;
+pub const OP_SW: InstructionOpcode = 64;
+pub const OP_VERSION1_JALR: InstructionOpcode = 65;
```
Comment (doitian) [inline definitions/src/instructions.rs:90]: Any reason to allocate different instructon opcode? From the code base, it seems that `OP_VERSION1_JALR` and `OP_JALR` lead to identical logic.

Comment (xxuejie) [reply definitions/src/instructions.rs:90]: The 2 opcodes actually do use different logic:

https://github.com/nervosnetwork/ckb-vm/blob/01cf746f32af6be7be2b574b14fe660e63df4719/src/machine/asm/execute.S#L448-L456
https://github.com/nervosnetwork/ckb-vm/blob/01cf746f32af6be7be2b574b14fe660e63df4719/src/machine/asm/execute.S#L1093-L1101

Notice the first 4 lines in each version, they are in different orders. The idea is that version 0 can use the original `OP_JALR`, while version 1 and later uses `OP_VERSION1_JALR`

File: src/instructions/common.rs [hunk 2]
```diff
@@ -80,13 +81,26 @@ pub fn addiw<Mac: Machine>(
 // =======================
 // #  LOAD instructions  #
 // =======================
+fn check_load_boundary<R: Register>(version0: bool, address: &R, bytes: u64) -> Result<(), Error> {
+    if version0 {
+        let address = address.to_u64();
+        let end = address.checked_add(bytes).ok_or(Error::OutOfBound)?;
+        if end == RISCV_MAX_MEMORY as u64 {
```
Comment (mohanson) [inline src/instructions/common.rs:88]: I am curious why it is not `end >= RISCV_MAX_MEMORY`

Comment (xxuejie) [reply src/instructions/common.rs:88]: The `>` case will be covered by normal memory module bound checking code. This bug only occurs with the `==` case.

**Comment by mohanson** [comment]
If we use `develop` and `release` branches to ensure that the mainnet is not broken, do we still need VERSION0 and VERSION1?

**Comment by xxuejie** [comment]
> If we use `develop` and `release` branches to ensure that the mainnet is not broken, do we still need VERSION0 and VERSION1?

I think we do, based on my understanding of forks, assuming we do a fork update at block 3000000, we will need to use VERSION0 to validate blocks under 3000000 to preserve compatibility, and use VERSION1 to validate blocks after 3000000. That means we will need the concept of version even if we split a release branch.

**Comment by doitian** [comment]
> > If we use `develop` and `release` branches to ensure that the mainnet is not broken, do we still need VERSION0 and VERSION1?
> 
> I think we do, based on my understanding of forks, assuming we do a fork update at block 3000000, we will need to use VERSION0 to validate blocks under 3000000 to preserve compatibility, and use VERSION1 to validate blocks after 3000000. That means we will need the concept of version even if we split a release branch.

It’s likely that we’ll run v0 for codes stored in cells created before a height, and v1 afterwards. Genesis system scripts are exceptions.

**Review by mohanson — APPROVED** [review]
LGTM

**Comment by xxuejie** [comment]
> It’s likely that we’ll run v0 for codes stored in cells created before a height, and v1 afterwards. Genesis system scripts are exceptions.

This makes sense but I don't think we need to make exceptions for genesis scripts. They can be treated as scripts created before the fork, hence v0 will be used. Later if there is the need to upgrade them, they might be created into new cells, which can leverage v1 or newer versions.

**Pull request merged**